nomt/src/bitbox/mod.rs

```diff
@@ -295,6 +245,72 @@ impl DB {
     }
 }
 
+/// A utility for loading pages from bitbox.
+pub struct PageLoader {
+    shared: Arc<Shared>,
+    meta_map: ArcRwLockReadGuard<parking_lot::RawRwLock, MetaMap>,
+    io_handle: IoHandle,
+}
+
+impl PageLoader {
+    /// Create a new page loader.
+    pub fn new(db: &DB, io_handle: IoHandle) -> Self {
+        PageLoader {
+            shared: db.shared.clone(),
+            meta_map: RwLock::read_arc(&db.shared.meta_map),
+            io_handle,
+        }
+    }
+
+    /// Load a page, blocking the current thread. Fails if the I/O pool is down.
+    pub fn load_sync(
+        &self,
+        ht_fd: RawFd,
+        page_id: &PageId,
+    ) -> anyhow::Result<Option<(Vec<u8>, BucketIndex)>> {
+        let mut probe_seq = ProbeSequence::new(page_id, &self.meta_map);
+
+        let mut page_buffer = None;
+
+        loop {
+            match probe_seq.next(&self.meta_map) {
+                ProbeResult::Tombstone(_) => continue,
+                ProbeResult::PossibleHit(bucket) => {
+                    // if this could be a match we need to fetch the page and check its id
+                    let data_page_index = self.shared.store.data_page_index(bucket);
+                    let buffer = page_buffer.take().unwrap_or_else(|| Box::new(Page::zeroed()));
+
+                    // send the read command
+                    let command = IoCommand {
+                        kind: IoKind::Read(ht_fd, data_page_index, buffer),
+                        user_data: 0,
+                    };
+
+                    self.io_handle.send(command).map_err(|_| anyhow::anyhow!("I/O pool hangup"))?;
+                    let complete = self.io_handle.recv()?;
+
+                    complete.result?;
+
+                    match complete.command.kind {
+                        IoKind::Read(fd, page_index, buffer)
+                            if fd == ht_fd && page_index == data_page_index
+                        => {
+                            if buffer[PAGE_SIZE - 32..] == page_id.encode() {
+                                return Ok(Some((buffer.to_vec(), BucketIndex(bucket))));
+                            } else {
+                                // misprobe. continue
```

Comment:
Would it be a useful metric here to record how many misses there are?

Reply:
Totally, this would help to understand the read amplification. Although for a small ratio of allocated_pages to total_hash_table_buckets, we aim to avoid misprobes (thanks to triangular probing)